server/src/instant/grouped_queue.clj

```diff
@@ -0,0 +1,82 @@
+(ns instant.grouped-queue
+  (:refer-clojure :exclude [peek])
+  (:import
+   (java.util.concurrent LinkedBlockingQueue TimeUnit)
+   (java.util.concurrent.atomic AtomicInteger)
+   (clojure.lang PersistentQueue)))
+
+(def empty-q PersistentQueue/EMPTY)
+
+(defn create [{:keys [group-fn]}]
+  {:size (AtomicInteger. 0)
+   :group-fn group-fn
+   :group-key->subqueue (atom {})
+   :dispatch-queue (LinkedBlockingQueue.)})
+
+(defn size [{:keys [size] :as _grouped-q}]
+  (.get size))
+
+(defn put! [{:keys [group-fn dispatch-queue group-key->subqueue size]
+             :as _grouped-q} item]
+  (let [group-key (group-fn item)]
+    (if (nil? group-key)
+      ;; This item is not to be grouped.
+      (do (.incrementAndGet size)
+          (.put dispatch-queue [:item item]))
+
+      ;; This item will be grouped on `group-key` 
+      (let [_ (.incrementAndGet size)
+            [prev] (swap-vals! group-key->subqueue
+                               update
+                               group-key
+                               (fn [subqueue]
+                                 (conj (or subqueue empty-q)
+                                       item)))
+            prev-subqueue (get prev group-key)
+            first-enqueue? (empty? prev-subqueue)]
+        (when first-enqueue?
+          (.put dispatch-queue [:group-key group-key]))))))
+
+(defn peek [{:keys [dispatch-queue group-key->subqueue] :as _grouped-q}]
+  (let [[t arg :as entry] (.peek dispatch-queue)]
+    (cond
+      (nil? entry) nil
+      (= t :item) arg
+      (= t :group-key) (first (get @group-key->subqueue arg)))))
+
+(defn process-polling!
+  ([gq process-fn] (process-polling! gq process-fn {:poll-ms 1000}))
+  ([{:keys [dispatch-queue group-key->subqueue size] :as _grouped-q}
+    process-fn
+    {:keys [poll-ms]}]
+   (let [[t arg :as entry] (.poll dispatch-queue poll-ms TimeUnit/MILLISECONDS)]
```

Comment:
Why do we call `poll` here instead of `take`? Is it because we want to be able to handle shutdown?

Reply:
Yes! I wanted a way to be able to handle the shutdown. Open for other ways though